model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java

```diff
@@ -239,6 +241,52 @@ public void removeExpired(RealmModel realm) {
 
     }
 
+    @Override
+    public UserSessionModel loadUserSession(String userSessionId, boolean offline) {
+
+        String offlineStr = offlineToString(offline);
+
+        TypedQuery<PersistentUserSessionEntity> userSessionQuery = em.createNamedQuery("findUserSession", PersistentUserSessionEntity.class);
+        userSessionQuery.setParameter("offline", offlineStr);
+        userSessionQuery.setParameter("userSessionId", userSessionId);
+        userSessionQuery.setMaxResults(1);
+
+        List<PersistentUserSessionAdapter> result = userSessionQuery.getResultStream()
+                .map(this::toAdapter)
+                .collect(Collectors.toList());
+
+        if (result.isEmpty()) {
+            return null;
+        }
+
+        PersistentUserSessionAdapter userSession = result.get(0);
+
+        TypedQuery<PersistentClientSessionEntity> clientSessionQuery = em.createNamedQuery("findClientSessionsByUserSessions", PersistentClientSessionEntity.class);
+        clientSessionQuery.setParameter("userSessionIds", Collections.singleton(userSessionId));
+        clientSessionQuery.setParameter("offline", offlineStr);
+        List<PersistentClientSessionEntity> clientSessions = clientSessionQuery.getResultList();
```

Comment:
Also consider using streams here:
```suggestion
        closing(clientSessionQuery.getResultStream()).forEach(clientSession -> {
            PersistentAuthenticatedClientSessionAdapter clientSessAdapter = toAdapter(userSession.getRealm(), userSession, clientSession);
            Map<String, AuthenticatedClientSessionModel> currentClientSessions = userSession.getAuthenticatedClientSessions();

            // Case when client was removed in the meantime
            if (clientSessAdapter.getClient() == null) {
                removedClientUUIDs.add(clientSession.getClientId());
            } else {
                currentClientSessions.put(clientSession.getClientId(), clientSessAdapter);
            }
        });
```

Reply:
I adapted the closing(..) part, but kept the rest as is, see [my comment](https://github.com/keycloak/keycloak/pull/7722#issuecomment-763155458) below